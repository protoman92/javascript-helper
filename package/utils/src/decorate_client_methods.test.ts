import { Mapper } from "@haipham/javascript-helper-essential-types";
import decorateClientMethods from "./decorate_client_methods";

describe("Decorate client methods", () => {
  const methodDecorator: Mapper<any, any> = (fn) => {
    return function (...args: any[]) {
      // @ts-ignore
      return fn.call(this, ...args);
    };
  };

  it("Should decorate object instance methods correctly", async () => {
    // Setup
    function Client() {}

    Client.prototype.method1 = function () {
      return this;
    };

    Client.prototype.method2 = function () {
      return 2;
    };

    Object.defineProperty(Client.prototype, "property1", {
      get: function () {
        return this;
      },
    });

    Object.defineProperty(Client.prototype, "property2", {
      get: function () {
        return () => {
          return this;
        };
      },
    });

    Object.defineProperty(Client.prototype, "property3", {
      get: function () {
        return function () {
          // @ts-ignore
          return this;
        };
      },
    });

    // When
    // @ts-ignore
    const client = new Client();

    const clientClone = decorateClientMethods<any>({
      decorator: methodDecorator,
    })(client);

    // Then
    // @ts-ignore
    expect(clientClone.property1).toStrictEqual(client);
    // @ts-ignore
    expect(clientClone.property2()).toStrictEqual(client);
    // @ts-ignore
    expect(clientClone.property3()).toStrictEqual(clientClone);
    // @ts-ignore
    expect(clientClone.method1()).toStrictEqual(client);
    // @ts-ignore
    expect(clientClone.method2()).toStrictEqual(2);
    expect(clientClone).toMatchSnapshot();
  });

  it("Should decorate class instance methods correctly", async () => {
    // Setup
    class Client {
      get property1() {
        return this;
      }

      get property2() {
        return () => {
          return this;
        };
      }

      get property3() {
        return function () {
          // @ts-ignore
          return this;
        };
      }

      method1() {
        return this;
      }

      method2 = () => {
        return this;
      };
    }
    // When
    const client = new Client();

    const clientClone = decorateClientMethods<typeof client>({
      decorator: methodDecorator,
    })(client);

    // Then
    expect(clientClone.property1).toStrictEqual(client);
    expect(clientClone.property2()).toStrictEqual(client);
    expect(clientClone.property3()).toStrictEqual(clientClone);
    expect(clientClone.method1()).toStrictEqual(client);
    expect(clientClone.method2()).toStrictEqual(client);
    expect(clientClone).toMatchSnapshot();
  });

  it("Should decorate key-value object methods correctly", async () => {
    // Setup
    const client = {
      get property1() {
        return this;
      },
      get property2() {
        return () => {
          return this;
        };
      },
      get property3() {
        return function () {
          // @ts-ignore
          return this;
        };
      },
      method1: function () {
        return this;
      },
      method2: () => {
        return this;
      },
    };

    // When
    const clientClone = decorateClientMethods<typeof client>({
      decorator: methodDecorator,
    })(client);

    // Then
    expect(clientClone.property1).toStrictEqual(client);
    expect(clientClone.property2()).toStrictEqual(client);
    expect(clientClone.property3()).toStrictEqual(clientClone);
    expect(clientClone.method1()).toStrictEqual(client);
    expect(clientClone.method2()).toStrictEqual(this);
    expect(clientClone).toMatchSnapshot();
  });
});
