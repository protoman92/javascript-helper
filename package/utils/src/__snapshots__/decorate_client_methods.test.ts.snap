// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decorate client methods Should decorate class instance methods correctly 1`] = `
Object {
  "method2": [Function],
}
`;

exports[`Decorate client methods Should decorate key-value object methods correctly 1`] = `
Object {
  "method1": [Function],
  "method2": [Function],
  "property1": Object {
    "method1": [Function],
    "method2": [Function],
    "property1": [Circular],
    "property2": [Function],
    "property3": [Function],
  },
  "property2": [Function],
  "property3": [Function],
}
`;

exports[`Decorate client methods Should decorate object instance methods correctly 1`] = `
Object {
  "method1": [Function],
  "method2": [Function],
}
`;
